### PR TITLE
Re-factor list of YAML keys for hooks to late-render

### LINKED ISSUE #6411: [CT-1634] [Bug] Post-hook in dbt_project.yml using {{ this }} failing to insert table name
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

Adding a post-hook into the `dbt_project.yml` file does not appear to be resolving `{{ this }}` to a table reference. 

### Expected Behavior

I expect that the reference `{{ this }}` renders to the correct target.

### Steps To Reproduce

1. Set up new project with `dbt init`
2. Insert the following into the `dbt_profile.yml` file:
```
models:
  +post_hook: 
    - "ALTER TABLE {{ this }} ADD COLUMN updated_at TIMESTAMP DEFAULT now()"
  test:
    # Config indicated by + and applies to all files under models/example/
    example:
      +materialized: view
```

### Relevant log output

```shell
%: dbt run
21:19:32  Running with dbt=1.3.1
21:19:32  Unable to do partial parsing because a project config has changed
21:19:32  Found 2 models, 4 tests, 0 snapshots, 0 analyses, 289 macros, 0 operations, 0 seed files, 0 sources, 0 exposures, 0 metrics
21:19:32  
21:19:33  Concurrency: 4 threads (target='dev')
21:19:33  
21:19:33  1 of 2 START sql table model test.my_first_dbt_model ........................... [RUN]
21:19:34  1 of 2 ERROR creating sql table model test.my_first_dbt_model .................. [ERROR in 0.51s]
21:19:34  2 of 2 SKIP relation test.my_second_dbt_model .................................. [SKIP]
21:19:34  
21:19:34  Finished running 1 table model, 1 view model in 0 hours 0 minutes and 2.19 seconds (2.19s).
21:19:34  
21:19:34  Completed with 1 error and 0 warnings:
21:19:34  
21:19:34  Database Error in model my_first_dbt_model (models/example/my_first_dbt_model.sql)
21:19:34    syntax error at or near "COLUMN"
21:19:34    LINE 3:         ALTER TABLE  ADD COLUMN updated_at TIMESTAMP DEFAULT...
21:19:34                                     ^
21:19:34    compiled Code at target/run/test/models/example/my_first_dbt_model.sql
21:19:34  
21:19:34  Done. PASS=0 WARN=0 ERROR=1 SKIP=1 TOTAL=2
```

### Environment

```markdown
- OS: Mac OSX
- Python: 3.10.7
- dbt: 1.3.1
- adapter: PostGres
```

### Which database adapter are you using with dbt?

postgres

### Additional Context

I have tried to use the dictionary approach outlined here: https://docs.getdbt.com/reference/resource-configs/pre-hook-post-hook#dbt_projectyml-use-a-dictionary, but that results in a different error (not included here). 

When I put the post hook directly into a config macro in my models, it works fine. I'm trying to avoid copy/pasting that code across all my models, if I can help it.

Thank you!

### PR DESCRIPTION
**Pull request by dbeatty10**
resolves #6411

### Description

The documentation [here](https://docs.getdbt.com/reference/resource-configs/pre-hook-post-hook) appears accurate. But as a user, it's easy for me to miss the subtle difference of a dash and a hyphen and to notice when/where each is relevant.

Is it just possible to treat both `post-hook` and `post_hook` as being synonymous? This pull request hopes the answer is YES!

### Example

`models/my_simple_model.sql`
```
select 1 as "id"
```

Run it:
```shell
dbt run --select my_simple_model
```

😄  Happy:
`dbt_project.yml`
```yaml
models:
  +post-hook:
    - "select 'This works: {{ this }}'"
```

☹️  Sad:
`dbt_project.yml`
```yaml
models:
  +post_hook: 
    - "select 'But this does not: {{ this }}'"
```

### What this pull request needs
- [x] create a new test (or modify an existing one) to cover the sad case shown above
- [x] verify that test fails as-is
- [x] verify it succeeds after making a fix

### Approach to testing

Rather than being exhaustive with testing all the variations across models, snapshots, and tests, I opted for just extending the basic test case for models. This was to minimize code added to `tests/functional/hooks/test_model_hooks.py` as well as not excessively lengthening the time to run tests. If required, we can revisit this decision.

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: core/dbt/config/renderer.py [hunk 1]
```diff
@@ -159,7 +159,8 @@ def should_render_keypath(self, keypath: Keypath) -> bool:
         if first in {"seeds", "models", "snapshots", "tests"}:
             keypath_parts = {(k.lstrip("+ ") if isinstance(k, str) else k) for k in keypath}
             # model-level hooks
-            if "pre-hook" in keypath_parts or "post-hook" in keypath_parts:
+            late_rendered_hooks = ["pre-hook", "post-hook", "pre_hook", "post_hook"]
+            if any(hook in keypath_parts for hook in late_rendered_hooks):
```
[the remark this answers is not in the document]
Comment (dbeatty10) [reply core/dbt/config/renderer.py:163]: Thanks for the suggestion -- `intersection` it is!

**Comment by dbeatty10** [comment]
Manually confirmed this fixes the issue. Would be nice to have a test to cover this case to protect against regressions.

**Comment by dbeatty10** [comment]
> If there aren't existing tests you can piggyback off, happy to help you set up new tests.

That would be awesome if you could help with the testing part @aranke !

I don't yet know if there are any good existing tests to piggyback from, but the following might help search for candidates:
```shell
git grep "pre_hook" tests/
git grep "post_hook" tests/
git grep "pre-hook" tests/
git grep "post-hook" tests/
```

**Comment by jtcohen6** [comment]
Nice work on this @dbeatty10! Much better than just updating the docs :)

After a few minutes of quick looking, I wasn't able to find a good test for piggy-backing. There's [one test I could find](https://github.com/dbt-labs/dbt-core/blob/d43c070007b823588e7a170c6c37495ce319b011/test/unit/test_config.py#L660-L661) with "macro in a `p*-hook` config defined in `dbt_project.yml`, but it's a unit test that doesn't do any actual Jinja rendering. I know I always end up opting for functional tests, because they're easiest to write, but that might be appropriate here given the number of different parsing steps that are in the mix for this one.

**Comment by dbeatty10** [comment]
I've been struggling to think of a simple platform-independent approach to test that `post-hook` and `post_hook` are indeed synonymous.

My first thoughts were to add some SQL/DML/etc that modifies the database state, and then we could assert that the intended state is observed. But it's hard to think of a state-changing statement that can be executed in a cross-database manner.

As an alternative approach, what if we try specifying an invalid statement and assert that it raises an error? This would confirm that the hook is being executed regardless if it's specified by `post-hook` or `post_hook` -- if it doesn't raise an error, it means that the config didn't take effect (or somehow the statement was allowed by the data platform).

The project configs for the latter approach might look similar to this:
```yaml
models:
  +post-hook:
    - "This is an invalid statement and should raise an exception"
```
and
```yaml
models:
  +post_hook:
    - "This is an invalid statement and should also raise an exception"
```

**Comment by dbeatty10** [comment]
> I've been struggling to think of a simple platform-independent approach to test that `post-hook` and `post_hook` are indeed synonymous.

Turns out it doesn't need to be platform-independent -- guessing can just add/extend relevant functional tests [here](https://github.com/dbt-labs/dbt-core/blob/89cc89dfdf76b2cb2054daa56eaed317441ec000/tests/functional/hooks/test_model_hooks.py).

**Comment by dbeatty10** [comment]
Extended a functional test [here](https://github.com/dbt-labs/dbt-core/pull/6435/commits/e41861c43089da96912f4268f27473293df48636) and [verified that it fails](https://github.com/dbt-labs/dbt-core/actions/runs/3885041600/jobs/6628399971#step:8:1906) when the bugfix is [removed](https://github.com/dbt-labs/dbt-core/pull/6435/commits/e83a2be2eb6f0ef4d95ffcfaf77ea9287eb0fbca):
```
=========================== short test summary info ============================
ERROR tests/functional/hooks/test_model_hooks.py::TestPrePostModelHooksUnderscores::test_pre_and_post_run_hooks - dbt.exceptions.CompilationException: Compilation Error
```

**Pull request merged**